DoLah/Parser/DateTimeParser.cpp

```diff
@@ -116,56 +362,93 @@ namespace DoLah {
                 throw e;
             }
         }
+
         return output;
     }
 
+    bool DateTimeParser::isValidDate(std::tm date) {
+        std::tm checkTime = date;
+        std::mktime(&checkTime);
+        // IF mktime changes the time format, that means the input date was wrong!
+        // ex. 31st April becomes 1st May
+        if (date.tm_mday != checkTime.tm_mday
+            || date.tm_mon != checkTime.tm_mon
+            || date.tm_year != checkTime.tm_year) {
+            return false;
+        }
+        return true;
+    }
+
     std::tm DateTimeParser::checkDMYformat(std::vector<std::string> strArr) {
         time_t t = time(0);
-        std::tm output;
-        localtime_s(&output, &t);
+        std::tm current;
+        localtime_s(&current, &t);
+
+        std::tm output = current;
 
         int day = -1;
         int month = -1;
         int year = -1;
 
+        int monthModifier = 0;
+        int yearModifier = 0;
+
         size_t size = strArr.size();
         day = getDay(strArr.at(0));
         if (day != REJECT) {
             output.tm_mday = day;
-            if (size <= 1) {
-                return output;
-            }
-            month = getMonth(strArr.at(1));
-            if (month != REJECT) {
-                output.tm_mon = month;
-                if (size <= 2) {
-                    return output;
-                }
-                year = getYear(strArr.at(2));
-                if (year != REJECT) {
-                    output.tm_year = year - 1900;
-                    return output;
+            if (size > 1) {
+                month = getMonth(strArr.at(1));
+                if (month != REJECT) {
+                    output.tm_mon = month;
+                    if (size > 2) {
+                        year = getYear(strArr.at(2));
+                        if (year != REJECT) {
+                            output.tm_year = year - 1900;
+                        } else {
+                            throw std::invalid_argument("");
+                        }
+                    } else {
+                        if (current.tm_mon > month) {
+                            yearModifier += 1;
+                        }
+                    }
```

Comment:
One easier solution is to flatten the condition, spread out the nested condition into one or two level deep only. 

There are many other possible solution such as polymorphic. Or perhaps use a callback function in a hashmap.

As one quotes:
> Flat is better than nested.